File: Week 5/C++ Codes/ThreeSum.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
    public:
        int threeSumClosest(vector<int> &A, int B) {
        sort(A.begin(), A.end());

        int bestSum = INT_MAX;
        long long int sum = 0;

        // Fix the smallest number in the three integers
        for (int i = 0; i < A.size() - 2; i++) {
            // Now A[i] is the smallest number in the three integers in the solution
            int left = i + 1, right = A.size() - 1;

            while (left < right) {
                sum = A[i] + A[left] + A[right];
                if (abs(B - sum) < abs(B - bestSum))
                    bestSum = sum;
                if (sum > B)
                    right--;
                else
                    left++;
            }
        }
        
        return bestSum;
    }
};

File: Week 5/Python Codes/ThreeSum.py
class Solution:
    # @param A : list of integers
    # @param B : integer
    # @return an integer
    def threeSumClosest(self, A, B):
        A.sort()
        bestSum, sum = float('inf'), 0
        # Fix the smallest number in the three integers
        for i in range(0, len(A)-2):
            # Now A[i] is the smallest number in the three integers in the solution
            left, right = i + 1, len(A) - 1
            while (left < right):
                sum = A[i] + A[left] + A[right]
                if (abs(B - sum) < abs(B - bestSum)):
                    bestSum = sum
                if (sum > B):
                    right -= 1
                else:
                    left += 1
        return bestSum